env/kube_registry.go

```diff
@@ -0,0 +1,331 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package env
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"reflect"
+	"strings"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	multierror "github.com/hashicorp/go-multierror"
+
+	"istio.io/manager/model"
+
+	"k8s.io/client-go/1.5/kubernetes"
+	"k8s.io/client-go/1.5/pkg/api"
+	"k8s.io/client-go/1.5/pkg/api/errors"
+	"k8s.io/client-go/1.5/pkg/api/unversioned"
+	"k8s.io/client-go/1.5/pkg/api/v1"
+	"k8s.io/client-go/1.5/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/1.5/pkg/runtime"
+	"k8s.io/client-go/1.5/pkg/runtime/serializer"
+	"k8s.io/client-go/1.5/rest"
+	"k8s.io/client-go/1.5/tools/clientcmd"
+)
+
+const IstioAPIGroup = "istio.io"
+const IstioResourceVersion = "v1"
+
+// KubernetesRegistry bindings for the manager:
+// - configuration objects are stored as third-party resources
+type KubernetesRegistry struct {
+	mapping   model.KindMap
+	client    *kubernetes.Clientset
+	dyn       *rest.RESTClient
+	Namespace string
+}
+
+// CreateRESTConfig for cluster API server, pass empty config file for in-cluster
+func CreateRESTConfig(kubeconfig string, km model.KindMap) (*rest.Config, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	version := unversioned.GroupVersion{
+		Group:   IstioAPIGroup,
+		Version: IstioResourceVersion,
+	}
+
+	config.GroupVersion = &version
+	config.APIPath = "/apis"
```

Comment:
Maybe, but that constant should have been defined upstream in Kubernetes. The name seems to be required to be "apis".